torch/csrc/autograd/python_cpp_function.cpp

```diff
@@ -26,11 +26,17 @@ PyObject* THPCppFunction_call(PyObject* self, PyObject* args, PyObject *kwargs)
     return PyErr_Format(PyExc_TypeError, "keyword arguments are not supported");
   }
 
+  auto& fn = *((THPCppFunction*)self);
   int num_inputs = PyTuple_GET_SIZE(args);
+  if (num_inputs != fn.num_args) {
+    return PyErr_Format(PyExc_TypeError, "expect %d arguments (got %d)",
+        fn.num_args, num_inputs);
+  }
+
   variable_list vars(num_inputs);
   for (int i = 0; i != num_inputs; ++i) {
     PyObject* arg = PyTuple_GET_ITEM(args, i);
-    if (arg == Py_None) {
+    if (i >= fn.num_required_args && arg == Py_None) {
```

Comment:
We don't have any functions that allow `(arg1, None, arg2)`. Required args have to come before optional args.

Checking it at the function level is a problem because it's easy to miss as we add new ops.